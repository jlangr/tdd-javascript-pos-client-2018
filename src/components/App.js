import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import Item from './Item';

class App extends Component {
  constructor() {
    super();

    this.state = { items: []};
  }

  voidItem = id => {
    const items = this.state.items.filter(item => item.id !== id);
    this.setState({ items });
  }

  // class method--`this` automatically transferred to this (no need for bind!)
  addItem = () => {
    const { items } = this.state;

    const ids = this.state.items.map(item => item.id);
    const max_id = ids.length > 0 ? Math.max(...ids) : 0;

    items.push({ id: max_id  + 1 });

    this.setState({ items });
  };

  render() {
    return(
      <div>
        <h2>Welcome</h2>
        <div className = 'item-list'>
          {this.state.items.map(item => {
            return (
              <Item 
                key={item.id}
                item={item}
                voidItem={this.voidItem}></Item>
            );
          })}
        </div>
        <Button className='btn-add' onClick={this.addItem}>Add Item</Button>
      </div>
    );
  }
};

export default App;