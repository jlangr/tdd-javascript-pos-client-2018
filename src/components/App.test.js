import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

describe('App', () => {
  const app = shallow(<App />);
  const id = 1;

  it('renders correct', () => {
    expect(app).toMatchSnapshot();
  });

  it('initializes the `state` with an empty list of gifts', () => {
    expect(app.state().items).toEqual([]);
  });

  describe('when clicking `scan item`', () => {
    beforeEach(() => {
      app.setState({ items: [] });
      app.find('.btn-add').simulate('click');
    })

    it('adds a new item to `state`', () => {
      expect(app.state().items).toEqual([{ id }]);
    });

    it('adds a new item to rendered list', () => {
      expect(app.find('.item-list').children().length).toEqual(1);
    })

    it('creates an Item component', () => {
      expect(app.find('Item').exists()).toBe(true);
    });
  });

  describe('and the user wants to void item', () => {
    beforeEach(() => {
      app.setState({ items: [] });
    });

    it('removes the item from `state`', () => {
      app.find('.btn-add').simulate('click');

      app.instance().voidItem(id); // UGH id correlation

      expect(app.state().items).toEqual([]);     
    });
  });
});