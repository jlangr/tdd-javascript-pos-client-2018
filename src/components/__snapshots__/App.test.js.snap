// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Half Foods
</h2>,
        <hr />,
        <Connect(CustomerExperience) />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Half Foods",
        },
        "ref": null,
        "rendered": "Half Foods",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Half Foods
</h2>,
          <hr />,
          <Connect(CustomerExperience) />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Half Foods",
          },
          "ref": null,
          "rendered": "Half Foods",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": true,
    "hydrateIn": undefined,
  },
}
`;
